### PR TITLE
Fix #12254/#11773: Datatable move lazy check to last

### LINKED ISSUE #11773: DataTable: isLazy() evaluation error for rowExpansion nested table
### Describe the bug

When loading a nested table within the rowExpansion of an outer table, the code to dynamically determine if the inner table is lazy fails on a null value. This works on 13.0.8. Explicitly setting `lazy="false"` on the inner p:dataTable seems to alleviate the problem.

Stacktrace:
```
SEVERE: java.lang.IllegalArgumentException: No bean specified
        at org.apache.commons.beanutils.PropertyUtilsBean.getNestedProperty(PropertyUtilsBean.java:773)
        at org.apache.commons.beanutils.PropertyUtilsBean.getProperty(PropertyUtilsBean.java:885)
        at org.apache.commons.beanutils.PropertyUtils.getProperty(PropertyUtils.java:464)
        at org.apache.commons.beanutils.BeanComparator.compare(BeanComparator.java:163)
        at java.base/java.util.TreeMap.getEntryUsingComparator(TreeMap.java:374)
        at java.base/java.util.TreeMap.getEntry(TreeMap.java:344)
        at java.base/java.util.TreeMap.get(TreeMap.java:279)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at javax.el.BeanELResolver.invoke(BeanELResolver.java:155)
        at javax.el.CompositeELResolver.invoke(CompositeELResolver.java:79)
        at org.apache.el.parser.AstValue.getValue(AstValue.java:158)
        at org.apache.el.ValueExpressionImpl.getValue(ValueExpressionImpl.java:189)
        at org.apache.webbeans.el22.WrappedValueExpression.getValue(WrappedValueExpression.java:68)
        at com.sun.faces.facelets.el.TagValueExpression.getValue(TagValueExpression.java:73)
        at javax.faces.component.ComponentStateHelper.eval(ComponentStateHelper.java:194)
        at javax.faces.component.ComponentStateHelper.eval(ComponentStateHelper.java:181)
        at javax.faces.component.UIData.getValue(UIData.java:742)
        at org.primefaces.component.api.UIData.lambda$isLazy$0(UIData.java:87)
        at org.primefaces.util.ELUtils.getType(ELUtils.java:64)
        at org.primefaces.component.api.UIData.lambda$isLazy$1(UIData.java:85)
        at org.primefaces.util.ComponentUtils.eval(ComponentUtils.java:529)
        at org.primefaces.component.api.UIData.isLazy(UIData.java:81)
        at org.primefaces.component.datatable.DataTable.processEvent(DataTable.java:263)
        at com.sun.faces.lifecycle.RestoreViewPhase$1.visit(RestoreViewPhase.java:438)
        at com.sun.faces.component.visit.FullVisitContext.invokeVisitCallback(FullVisitContext.java:127)
        at org.primefaces.component.api.UIData.visitTree(UIData.java:671)
        at org.primefaces.component.api.UIData.visitColumnContent(UIData.java:887)
        at org.primefaces.component.api.UIData.visitRows(UIData.java:859)
        at org.primefaces.component.datatable.DataTable.visitRows(DataTable.java:763)
        at org.primefaces.component.api.UIData.visitTree(UIData.java:687)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1469)
        at javax.faces.component.UIForm.visitTree(UIForm.java:356)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1469)
        at javax.faces.component.UIComponent.visitTree(UIComponent.java:1469)
        at com.sun.faces.lifecycle.RestoreViewPhase.deliverPostRestoreStateEvent(RestoreViewPhase.java:433)
        at com.sun.faces.lifecycle.RestoreViewPhase.execute(RestoreViewPhase.java:257)
        at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:76)
        at com.sun.faces.lifecycle.RestoreViewPhase.doPhase(RestoreViewPhase.java:110)
        at com.sun.faces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:177)
        at javax.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:707)
        at javax.faces.webapp.FacesServlet.service(FacesServlet.java:451)
        at org.eclipse.jetty.servlet.ServletHolder$NotAsync.service(ServletHolder.java:1450)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
        at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
        at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:292)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:600)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:137)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
        at java.base/java.lang.Thread.run(Thread.java:1589)
```

### Reproducer

[rowexpand-el-error.zip](https://github.com/primefaces/primefaces/files/15028825/rowexpand-el-error.zip)

Steps:
1) Start reproducer with `mvn clean jetty:run`
2) Navigate to `localhost:8080`
3) Expand any row

### Expected behavior

Same code works in 13.0.8. 

### PrimeFaces edition

Community

### PrimeFaces version

14.0.0-RC3

### Theme

N/A

### JSF implementation

Mojarra

### JSF version

2.3

### Java version

JDK 11

### Browser(s)

All

### LINKED ISSUE #12254: Datatable: Nested datatable with rowExpansion calls backing bean at restore view phase with Primefaces 14 not fired with Primefaces 13
### Describe the bug

Primefaces 14 calls the WelcomeMBean method at the Restore View Phase, not with Primefaces 13.

I created a simple example to show the different behaviors.

### Reproducer

The example reproducer is at https://github.com/persapiens/primefaces-14
[README](https://github.com/persapiens/primefaces-14/blob/main/README.md) describes the steps to see the issue.

### Expected behavior

Using the reproducer provided and clicking the 'Submit' button on the welcome page,  I think it can be an issue with PrimeFaces 14 because the getDetails method of WelcomeMBean should not be called with an empty value at the Restore View Faces Phase.

I look forward to your answer.

### PrimeFaces edition

Community

### PrimeFaces version

14.0.2

### Theme

vela

### JSF implementation

Mojarra

### JSF version

4.0.6

### Java version

21

### Browser(s)

Chrome 126.0.6478.126 (Official Build) (64-bit)

### PR DESCRIPTION
**Pull request by melloware**
Fix #12254: Datatable move lazy check to last
Fix #11773: Datatable move lazy check to last

### CODE REVIEW

**Comment by Rapster** [comment]
Don't think it fixes the issue... If you enable filtering, and have active filters, you'll end up with the same issue no?

**Comment by melloware** [comment]
Yes but this was how it was before the change in 14. Basically lazy is the most expensive check now and I tested both tickets and they are back to working like they were in 13

So it might not totally fix it but it fixes these two issues specifically and restores pre-14 behavior.

**Comment by melloware** [comment]
Actually @Rapster both issues arose from row expansion and now it's evaluating the "value" of the datatable inside the expansion when it didn't before. So there is no way filtering would be active in this scenario.

**Comment by tandraschko** [comment]
can we add a integration test for it, not breaking it again in future?

**Comment by melloware** [comment]
Integration test added.  I have verified if this fix gets reverted this test fails because it blows up.

```java
throw new FacesException("GitHub #12254 was reverted as this should not throw an error on AJAX update!");
```

**Comment by Rapster** [comment]
Ok, so from what I understand during RESTORE_VIEW phase, SKIP_ITERATION is set (by both impl I imagine) therefore the var bean will never be set and as a consequence you'll get a NPE. The best we can do is this PR, but eventually the problem will occur if the nested table enable filtering... In this case, no other choice than setting the lazy attr with a true/false

**Review by Rapster — APPROVED** [review]

**Comment by melloware** [comment]
@Rapster you have assessed it accurately.  It could be combined with my other closed PR where we catch the exception and log them a better WARNING as to what to do?

**Comment by Rapster** [comment]
I take back what I said, RowExpansion might be the issue... I was wondering how come we end up parsing the EL so late since it should have been in the viewstate, but since the rowexpansion is not expanded therefore not encoded, then the nested lazy EL is not parsed (not in viewstate then). If you set `DataTable#expandedRow` (which is a strange to be TBH) it works, because it's been encoded...

Anyway, even though we have a better understanding of whats going on... I don't have a better fix than this one yet... We can have specific log for this scenario (so people don't raise an issue everytime they face that problem)

**Comment by melloware** [comment]
OK I added the explicit error checking back to `isLazy()` and suggested what the developer can do to mitigate the issue.

```java
 LOGGER.warning("Unable to automatically determine the `lazy` attribute, fallback to false. "
         + "Either define the `lazy` attribute on the component or make sure the `value` attribute doesn't resolve to `null`. "
          + "clientId: " + this.getClientId());
```

**Comment by Rapster** [comment]
The issue that I have with this log, is that, in dev mode (as it should be quite verbose) people might not see it and will take no actions. I'm cool with the first one (and yet, don't know in which scenario it's not able to determine the type but it's quite surprising...) but second one should be in error and whatever the project stage

**Comment by melloware** [comment]
OK updated to an ERROR an always regardless of ProjectStage.

File: primefaces/src/main/java/org/primefaces/component/api/UIData.java [hunk 1]
```diff
@@ -80,18 +82,27 @@ public enum PropertyKeys {
...
-                        + "clientId: " + this.getClientId());
+            try {
+                // if not set by xhtml, we need to check the type of the value binding
+                Class<?> type = ELUtils.getType(context, getValueExpression("value"), () -> getValue());
+
+                if (type == null) {
+                    if (LOGGER.isLoggable(Level.WARNING) && context.isProjectStage(ProjectStage.Development)) {
+                        LOGGER.warning("Unable to automatically determine the `lazy` attribute, fallback to false. "
+                                + "Either define the `lazy` attribute on the component or make sure the `value` attribute doesn't resolve to `null`. "
+                                + "clientId: " + this.getClientId());
```
Comment (Rapster) [inline primefaces/src/main/java/org/primefaces/component/api/UIData.java]: remove useless this

File: primefaces/src/main/java/org/primefaces/component/api/UIData.java [hunk 2]
```diff
@@ -80,18 +82,27 @@ public enum PropertyKeys {
...
+                else {
+                    lazy = LazyDataModel.class.isAssignableFrom(type);
+                }
             }
-            else {
-                lazy = LazyDataModel.class.isAssignableFrom(type);
+            catch (Exception e) {
+                LOGGER.severe("Exception occurred while determining the `lazy` attribute, fallback to false. "
+                            + "To prevent this error set the `lazy` property directly on the component. "
+                            + "Error: " + e.getMessage() + ". clientId: " + this.getClientId());
```
Comment (Rapster) [inline primefaces/src/main/java/org/primefaces/component/api/UIData.java]: remove useless this

**Review by Rapster — CHANGES_REQUESTED** [review]
last coment and all good

File: primefaces/src/main/java/org/primefaces/component/api/UIData.java (hunk 2 above)
Comment (melloware) [reply primefaces/src/main/java/org/primefaces/component/api/UIData.java]: Done!

**Pull request merged**